From ed9d9fc7cef65530e859082125ae91efbab61780 Mon Sep 17 00:00:00 2001
From: Ban Feng <kcfeng0@nuvoton.com>
Date: Fri, 7 Apr 2023 09:21:44 +0800
Subject: [PATCH] psusensor: rename device type IPSPS to IPSPS1

Signed-off-by: Ban Feng <kcfeng0@nuvoton.com>
---
 src/PSUSensorMain.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/PSUSensorMain.cpp b/src/PSUSensorMain.cpp
index 07802cd4..18b2b6eb 100644
--- a/src/PSUSensorMain.cpp
+++ b/src/PSUSensorMain.cpp
@@ -53,7 +53,7 @@ static const I2CDeviceTypeMap sensorTypes{
     {"DPS800", I2CDeviceType{"dps800", true}},
     {"INA219", I2CDeviceType{"ina219", true}},
     {"INA230", I2CDeviceType{"ina230", true}},
-    {"IPSPS", I2CDeviceType{"ipsps", true}},
+    {"IPSPS1", I2CDeviceType{"ipsps1", true}},
     {"IR38060", I2CDeviceType{"ir38060", true}},
     {"IR38164", I2CDeviceType{"ir38164", true}},
     {"IR38263", I2CDeviceType{"ir38263", true}},
-- 
2.34.1

