From 7721070d2187d0d8536879f30099fd8d0fb3672c Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Tue, 19 Nov 2024 09:23:38 +0800
Subject: [PATCH] hwmon temp: add lm75 support

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 src/hwmon-temp/HwmonTempMain.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/hwmon-temp/HwmonTempMain.cpp b/src/hwmon-temp/HwmonTempMain.cpp
index a9ec2c13..0a9ffc49 100644
--- a/src/hwmon-temp/HwmonTempMain.cpp
+++ b/src/hwmon-temp/HwmonTempMain.cpp
@@ -63,6 +63,7 @@ static constexpr double maxValueTemperature = 127;      // DegreesC
 static constexpr double minValueTemperature = -128;     // DegreesC
 
 static const I2CDeviceTypeMap sensorTypes{
+    {"LM75", I2CDeviceType{"lm75", true}},
     {"ADM1021", I2CDeviceType{"adm1021", true}},
     {"DPS310", I2CDeviceType{"dps310", false}},
     {"EMC1403", I2CDeviceType{"emc1403", true}},
-- 
2.43.0

