From 18c82521deac72f772b903bf472cc96c6b0a8854 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Tue, 19 Nov 2024 09:23:38 +0800
Subject: [PATCH] hwmon temp: add lm75 support

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 src/hwmon-temp/HwmonTempMain.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/hwmon-temp/HwmonTempMain.cpp b/src/hwmon-temp/HwmonTempMain.cpp
index f7ccb4f0..edd728cf 100644
--- a/src/hwmon-temp/HwmonTempMain.cpp
+++ b/src/hwmon-temp/HwmonTempMain.cpp
@@ -65,6 +65,7 @@ static constexpr double minValueTemperature = -128;     // DegreesC
 namespace fs = std::filesystem;
 
 static const I2CDeviceTypeMap sensorTypes{
+    {"LM75", I2CDeviceType{"lm75", true}},
     {"ADM1021", I2CDeviceType{"adm1021", true}},
     {"DPS310", I2CDeviceType{"dps310", false}},
     {"EMC1403", I2CDeviceType{"emc1403", true}},
-- 
2.43.0

