From 149d55fb9a11be9d794a350945f89ee62eed6b9f Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Tue, 16 Aug 2022 16:00:00 +0800
Subject: [PATCH 08/18] Add reset SEL

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 include/dbus-sdr/sdrutils.hpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/include/dbus-sdr/sdrutils.hpp b/include/dbus-sdr/sdrutils.hpp
index a38673e6..2797b77c 100644
--- a/include/dbus-sdr/sdrutils.hpp
+++ b/include/dbus-sdr/sdrutils.hpp
@@ -317,13 +317,15 @@ enum class SensorTypeCodes : uint8_t
     buttons = 0x14,
     watchdog2 = 0x23,
     entity = 0x25,
+    subsystem_health = 0xE0,
 };
 
 enum class SensorEventTypeCodes : uint8_t
 {
     unspecified = 0x00,
     threshold = 0x01,
-    sensorSpecified = 0x6f
+    sensorSpecified = 0x6f,
+    oem = 0x70
 };
 
 const static boost::container::flat_map<
@@ -355,6 +357,8 @@ const static boost::container::flat_map<
                                      SensorEventTypeCodes::sensorSpecified)},
          {"entity", std::make_pair(SensorTypeCodes::entity,
                                    SensorEventTypeCodes::sensorSpecified)},
+         {"oem_health", std::make_pair(SensorTypeCodes::subsystem_health,
+                                       SensorEventTypeCodes::oem)},
          {"energy", std::make_pair(SensorTypeCodes::other,
                                    SensorEventTypeCodes::threshold)}}};
 std::string getSensorTypeStringFromPath(const std::string& path);
-- 
2.34.1

