From 79a211ca53e39ee091f5b1756c2918c45ba2b2f7 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Tue, 16 Aug 2022 16:23:46 +0800
Subject: [PATCH 12/18] sensor reading optional zero

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 dbus-sdr/sensorcommands.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dbus-sdr/sensorcommands.cpp b/dbus-sdr/sensorcommands.cpp
index a7308bc0..b8d8779f 100644
--- a/dbus-sdr/sensorcommands.cpp
+++ b/dbus-sdr/sensorcommands.cpp
@@ -1096,7 +1096,7 @@ ipmi::RspType<uint8_t, uint8_t, uint8_t, std::optional<uint8_t>>
     }
 
     // no discrete as of today so optional byte is never returned
-    return ipmi::responseSuccess(value, operation, thresholds, std::nullopt);
+    return ipmi::responseSuccess(value, operation, thresholds, (uint8_t)0);
 }
 
 /** @brief implements the Set Sensor threshold command
-- 
2.34.1

