From 1bebe4ee64ce52c186dd645bf123145b3d8c771c Mon Sep 17 00:00:00 2001
From: Brian Ma <chma0@nuvoton.com>
Date: Thu, 17 Feb 2022 08:55:45 +0800
Subject: [PATCH] Add support report same version error

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Brian Ma <chma0@nuvoton.com>
Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 image_manager.cpp | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/image_manager.cpp b/image_manager.cpp
index 46f023a..cafd4a2 100644
--- a/image_manager.cpp
+++ b/image_manager.cpp
@@ -15,6 +15,7 @@
 #include <phosphor-logging/elog.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <xyz/openbmc_project/Software/Image/error.hpp>
+#include <xyz/openbmc_project/Software/Version/error.hpp>
 
 #include <algorithm>
 #include <cstring>
@@ -34,11 +35,13 @@ namespace manager
 PHOSPHOR_LOG2_USING;
 using namespace phosphor::logging;
 using namespace sdbusplus::error::xyz::openbmc_project::software::image;
+using namespace sdbusplus::error::xyz::openbmc_project::software::version;
 namespace Software = phosphor::logging::xyz::openbmc_project::software;
 using ManifestFail = Software::image::ManifestFileFailure;
 using UnTarFail = Software::image::UnTarFailure;
 using InternalFail = Software::image::InternalFailure;
 using ImageFail = Software::image::ImageFailure;
+using ExitFail = Software::version::AlreadyExists;
 namespace fs = std::filesystem;
 
 struct RemovablePath
@@ -229,8 +232,10 @@ int Manager::processImage(const std::string& tarFilePath)
     }
     else
     {
-        info("Software Object with the same version ({VERSION}) already exists",
-             "VERSION", id);
+        error("Software Object with the same version ({VERSION}) already exists",
+              "VERSION", id);
+        report<AlreadyExists>(
+               ExitFail::IMAGE_VERSION(version.c_str()));
     }
     return 0;
 }
-- 
2.34.1

