From 2b99e882308bf4c31d334c6383620689938b107b Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Mon, 3 Feb 2025 11:51:25 +0800
Subject: [PATCH] disable mmc mirroruboot

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Brian Ma <chma0@nuvoton.com>
Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 bmc/obmc-flash-bmc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bmc/obmc-flash-bmc b/bmc/obmc-flash-bmc
index f9cc4e25..c72bece3 100644
--- a/bmc/obmc-flash-bmc
+++ b/bmc/obmc-flash-bmc
@@ -646,6 +646,7 @@ function mmc_setprimary() {
 }
 
 function mmc_mirroruboot() {
+    return
     # Get current boot device; 0-primary_bootdev device; 1 - alt_bootdev
     bootdev=$(cat /sys/kernel/debug/aspeed/sbc/abr_image)
     if [[ "${bootdev}" == "0" ]]; then
-- 
2.43.0

