From 3e400080e62c9e6620ff6cec8676dcbf5d535266 Mon Sep 17 00:00:00 2001
From: Ban Feng <kcfeng0@nuvoton.com>
Date: Tue, 6 Jun 2023 11:33:34 +0800
Subject: [PATCH] Create Associations From Defined Property

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Ban Feng <kcfeng0@nuvoton.com>
---
 src/entity_manager/entity_manager.cpp | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

Index: git/src/entity_manager/entity_manager.cpp
===================================================================
--- git.orig/src/entity_manager/entity_manager.cpp
+++ git/src/entity_manager/entity_manager.cpp
@@ -633,6 +633,34 @@ void postToDbus(const nlohmann::json& ne
             }
         }
 
+        auto findAssociations = boardValues.find("Associations");
+        if (findAssociations != boardValues.end() &&
+            findAssociations->type() == nlohmann::json::value_t::array)
+        {
+            std::cerr << "create Associations to " << boardNameOrig << "\n";
+            std::vector<Association> associations;
+            for (auto& association : *findAssociations)
+            {
+                auto findForward = association.find("Forward");
+                auto findReverse = association.find("Reverse");
+                auto findPath = association.find("Path");
+                if (findForward != association.end() &&
+                    findReverse != association.end() &&
+                    findPath != association.end())
+                {
+                    associations.emplace_back(
+                        findForward->get<std::string>(),
+                        findReverse->get<std::string>(),
+                        findPath->get<std::string>());
+                }
+            }
+             std::shared_ptr<sdbusplus::asio::dbus_interface> ifacePtr = dbus_interface::createInterface(
+                objServer, boardPath,
+		"xyz.openbmc_project.Association.Definitions", boardNameOrig);
+            ifacePtr->register_property("Associations", associations);
+            ifacePtr->initialize();
+        }
+
         auto exposes = boardValues.find("Exposes");
         if (exposes == boardValues.end())
         {
