From 58fe21afa4cbcbbcf87fbe3af82c7118ec9a686b Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Mon, 15 Aug 2022 10:52:31 +0800
Subject: [PATCH 1/2] skip POWER BUTTON

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 src/power_control.cpp | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/power_control.cpp b/src/power_control.cpp
index a4ef078..f4078de 100644
--- a/src/power_control.cpp
+++ b/src/power_control.cpp
@@ -766,12 +766,14 @@ static void powerRestorePolicyLog()
                     "OpenBMC.0.1.PowerRestorePolicyApplied", NULL);
 }
 
+#if 0
 static void powerButtonPressLog()
 {
     sd_journal_send("MESSAGE=PowerControl: power button pressed", "PRIORITY=%i",
                     LOG_INFO, "REDFISH_MESSAGE_ID=%s",
                     "OpenBMC.0.1.PowerButtonPressed", NULL);
 }
+#endif
 
 static void resetButtonPressLog()
 {
@@ -2042,6 +2044,7 @@ static void sioS5Handler(bool state)
     sendPowerControlEvent(powerControlEvent);
 }
 
+#if 0
 static void powerButtonHandler(bool state)
 {
     bool asserted = state == powerButtonConfig.polarity;
@@ -2060,6 +2063,7 @@ static void powerButtonHandler(bool state)
         }
     }
 }
+#endif
 
 static void resetButtonHandler(bool state)
 {
@@ -2778,7 +2782,7 @@ int main(int argc, char* argv[])
             return -1;
         }
     }
-
+#if 0
     // Request POWER_BUTTON GPIO events
     if (powerButtonConfig.type == ConfigType::GPIO)
     {
@@ -2794,7 +2798,7 @@ int main(int argc, char* argv[])
             power_control::dbusGPIOMatcher(powerButtonConfig,
                                            powerButtonHandler);
     }
-
+#endif
     // Request RESET_BUTTON GPIO events
     if (resetButtonConfig.type == ConfigType::GPIO)
     {
@@ -3213,6 +3217,7 @@ int main(int argc, char* argv[])
     sdbusplus::asio::object_server buttonsServer =
         sdbusplus::asio::object_server(conn);
 
+#if 0
     if (!powerButtonConfig.lineName.empty())
     {
         // Power Button Interface
@@ -3266,7 +3271,7 @@ int main(int argc, char* argv[])
 
         powerButtonIface->initialize();
     }
-
+#endif
     if (!resetButtonConfig.lineName.empty())
     {
         // Reset Button Interface
-- 
2.34.1

