From 697dc6cde72c65079c676de08f3bbe4aa32bf813 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Tue, 4 Jul 2023 17:22:44 +0800
Subject: [PATCH 5/5] add psu sesnor: p2011 and mx16550

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 src/PSUSensorMain.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/PSUSensorMain.cpp b/src/PSUSensorMain.cpp
index 74918109..71eb393c 100644
--- a/src/PSUSensorMain.cpp
+++ b/src/PSUSensorMain.cpp
@@ -43,6 +43,8 @@
 static constexpr bool debug = false;
 
 static const I2CDeviceTypeMap sensorTypes{
+    {"P2011", I2CDeviceType{"p2011", true}},
+    {"MAX16550", I2CDeviceType{"max16550", true}},
     {"ADC128D818", I2CDeviceType{"adc128d818", true}},
     {"ADM1266", I2CDeviceType{"adm1266", true}},
     {"ADM1272", I2CDeviceType{"adm1272", true}},
-- 
2.34.1

